massa-bootstrap/src/server.rs

```diff
@@ -14,365 +42,508 @@ use massa_signature::KeyPair;
 use massa_time::MassaTime;
 use parking_lot::RwLock;
 use std::{
-    collections::{hash_map, HashMap, HashSet},
-    io,
+    collections::HashMap,
     net::{IpAddr, SocketAddr},
-    path::PathBuf,
     sync::Arc,
+    thread,
     time::{Duration, Instant},
 };
-use tokio::{sync::mpsc, task::JoinHandle};
-use tracing::{debug, info, warn};
+use tokio::runtime::{self, Handle, Runtime};
+use tracing::{debug, error, info, warn};
 
 use crate::{
     error::BootstrapError,
     messages::{BootstrapClientMessage, BootstrapServerMessage},
     server_binder::BootstrapServerBinder,
-    tools::normalize_ip,
+    types::{Duplex, Listener},
     BootstrapConfig, Establisher,
 };
 
+/// Abstraction layer over data produced by the listener, and transported
+/// over to the worker via a channel
+type BsConn = (Duplex, SocketAddr);
 /// handle on the bootstrap server
 pub struct BootstrapManager {
-    join_handle: JoinHandle<Result<(), BootstrapError>>,
-    manager_tx: mpsc::Sender<()>,
+    update_handle: std::thread::JoinHandle<Result<(), Box<BootstrapError>>>,
```

Comment:
```suggestion
    update_handle: std::thread::JoinHandle<Result<(), Box<BootstrapError>>>,
    // need to preserve the listener handle up to here to prevent it being destroyed
```